Add release note for module level public and complex variables (#102)
ballerina-platform/ballerina-release: release-notes/swan-lake-alpha3-release-note.md

#### Language

> ##### Support for Module level complex variables

Reviewer: ```suggestion
##### Support for Module-level Variables with List, Mapping, and Error Binding Patterns
```

Variable declarations with list, mapping, and error binding patterns are now allowed at module level. Unlike simple variables, these variables must be initialized in the declaration.

> Complex variable declarations cannot contain `isolated` or `configurable` qualifier.

Reviewer: ```suggestion
Complex variable declarations cannot contain `isolated` or `configurable` qualifiers.
```

Reviewer: Or in-line with what @praneesha suggested in https://github.com/ballerina-platform/ballerina-release/pull/102#discussion_r589303917

```suggestion
Complex variable declarations cannot contain an `isolated` or `configurable` qualifier.
```

```ballerina
// Module-level variable declaration with a list binding pattern. 
[int, float] [a, b] = [1, 2.5];

// Module-level variable declaration with a mapping binding pattern.

> Person {name: firstname, married: isMarried} = {name: "Jhone", married: true};

Reviewer: Shall we use "John" instead?
```suggestion
Person {name: firstname, married: isMarried} = {name: "John", married: true};
```

type Person record {|
    string name;
    boolean married;
|};

// Module-level variable declaration with an error binding pattern.

> error error(message) = error("error message");

Reviewer: For the expressions on the RHS, shall we use function calls like you had done for error earlier? It may not be obvious to users the advantages of binding patterns when we create a new value just for this.

For the error binding pattern, let's also bind a detail field.

#### Language

##### Module level complex variable support

> Tuple variable, Record variable and Error variable declaration is now allowed at module level. Unlike simple variables, complex variable declaration must have an initialization.

Reviewer: ```suggestion
Variable declarations with list, mapping, and error binding patterns are now allowed at module level. Unlike simple variables, these variables must be initialized in the declaration.
```

> Complex variable declaration cannot contain `isolated` or `configurable` qualifier.

Reviewer: ```suggestion

Complex variable declarations cannot contain `isolated` or `configurable` qualifiers.

```

```ballerina

> // Module level tuple variable declaration

Reviewer: ```suggestion
// Module-level variable declaration with a list binding pattern.
```

[int, float] [a, b] = [1, 2.5];

> // Module level record variable declaration

Reviewer: ```suggestion
// Module-level variable declaration with a mapping binding pattern.
```

> Person {name: Fname, married: Married} = {name: "Jhone", married: true};

Reviewer: ```suggestion
Person {name: firstName, married: isMarried} = {name: "John", married: true};
```

> type Person record {|

Reviewer: Shall we move this definition above the variable declaration that uses `Person`?

Author: I moved it down to co-occur the comment and the declaration.

Reviewer: We can move this before L55, right?

    string name;
    boolean married;
|};

> // Module level error variable declaration

Reviewer: ```suggestion
// Module-level variable declaration with an error binding pattern.
```

> UserDefinedError error(message, basicErrorNo = detail) = getError();

Reviewer: Do we need a custom error? Shall we use the generic error?

type UserDefinedError error <BasicErrorDetail>;
type BasicErrorDetail record {|
    int basicErrorNo?;
    anydata...;

> |};

Reviewer: Shall we move these also to before where they are used?

Author: removed custom error

> type BasicErrorDetail record {|
>     int basicErrorNo?;
>     anydata...;
> |};

Reviewer: ```suggestion
type BasicErrorDetail record {
    int basicErrorNo?;
};
```

If we are using this, maybe we can have a field with a simple name like `code`?

Author: removed custom error

function getError() returns UserDefinedError {
    return error UserDefinedError("error message", basicErrorNo = 1);
}
```

> ##### Module level public variable support

Reviewer: ```suggestion
##### Support for Module-level Public Variables
```

> Module level variable can be declared as public using `public` qualifier which will be visible to any module. Isolated variables and variables declared with `var` cannot contain `public` qualifier.

Reviewer: ```suggestion
Module-level variables can now be declared as public using the `public` qualifier. Such variables will be visible outside the modules in which they are declared. 

Isolated variables and variables declared with `var` cannot be declared as public variables.

```

#### Language

> ##### Support for Module level Variables with List, Mapping, and Error Binding Patterns

Reviewer: ```suggestion
##### Support for Module-level Variables with List, Mapping, and Error Binding Patterns
```

#### Language

##### Support for Module level Variables with List, Mapping, and Error Binding Patterns

Variable declarations with list, mapping, and error binding patterns are now allowed at module level. Unlike simple variables, these variables must be initialized in the declaration.

Also, these variable declarations cannot contain an `isolated` or `configurable` qualifier.

```ballerina
type Person record {|
    string name;
    boolean married;
|};

> function getList() returns [int, float] {
>     return [1, 2.5];
> }

Reviewer: Let's use expression-bodied functions.

```suggestion
function getList() returns [int, float] => [1, 2.5];
```

> function getPerson() returns Person {
>     return {name: "John", married: true};
> }

Reviewer: ```suggestion
function getPerson() returns Person => {name: "John", married: true};
```

> function getError() returns error {
>     return error("error message");
> }

Reviewer: Let's also add a code detail field.
```suggestion
function getError() returns error => error("error message", code = 1001, fatal = true);
```

// Module-level variable declaration with a list binding pattern. 
[int, float] [a, b] = getList();

// Module-level variable declaration with a mapping binding pattern.
Person {name: firstName, married: isMarried} = getPerson();

// Module-level variable declaration with an error binding pattern.
error error(message, code = errCode) = getError();
```

##### Support for Module-level Public Variables

Module-level variables can now be declared as public using the `public` qualifier. Such variables will be visible outside the modules in which they are declared.

Isolated variables and variables declared with `var` cannot be declared as public variables.

```ballerina

> public string publicName = "Ballerina";

Reviewer: ```suggestion
public string name = "Ballerina";
```
